src/canvas.jl

```diff
@@ -74,24 +138,32 @@ function lines!(
     nsteps = abs(Δx) > abs(Δy) ? abs(Δx) : abs(Δy)
     nsteps = min(nsteps, typemax(Int32))  # hard limit
 
-    δx = Δx / nsteps
-    δy = Δy / nsteps
-
-    px, Px = extrema([x2p(mx), x2p(Mx)])
-    py, Py = extrema([y2p(my), y2p(My)])
+    px, Px = x2p(mx), x2p(Mx)
+    py, Py = y2p(my), y2p(My)
 
-    pixel!(c, floor(Int, cur_x), floor(Int, cur_y), color)
     max_num_iter = typemax(Int16)  # performance limit
-    for _ in if nsteps > max_num_iter
-        range(1, stop = nsteps, length = max_num_iter)
+    rng = if nsteps > max_num_iter
+        nsteps, max_num_iter, nothing
     else
-        range(1, stop = nsteps, step = 1)
+        nsteps, nothing, 1
     end
-        cur_x += δx
-        cur_y += δy
-        (cur_y < py || cur_y > Py) && continue
-        (cur_x < px || cur_x > Px) && continue
-        pixel!(c, floor(Int, cur_x), floor(Int, cur_y), color)
+
+    args = (
+        rng...,
+        min(px, Px),
+        max(px, Px),
+        min(py, Py),
+        max(py, Py),
+        cur_x,
+        cur_y,
+        Δx / nsteps,
+        Δy / nsteps,
+    )
+
+    if c_or_v1 isa AbstractFloat && c_or_v2 isa AbstractFloat && col_cb !== nothing
+        interp_color_line!(c, args..., 1 / √(Δx^2 + Δy^2), c_or_v1, c_or_v2, col_cb)
```

Comment:
Is this equivalent to the following?

```julia
x0 = SVector(min(px, Px), min(py, Py))
x1 = SVector(max(px, Px), max(py, Py))
r = LinRange(x0, x1, floor(Int, nsteps))
iΔ = 1 / √(Δx^2 + Δy^2)
for i in 1:length(r)
    w = sqrt(sum(abs2, x-x0)) * iΔ
    v = (1-w) * c_or_v1 + w * c_or_v2
    pixel!(c, floor.(Int, r[i])..., col_cb(v))
end
```

Reply:
Thanks, originally, I've split `interp_color_line!` and `const_color_line!` because of type stability issues. I'll see if this works out and simplifies.

Reply:
> ERROR: ArgumentError: range([0.0, 0.0], stop=[80.0, 60.0], length=1): endpoints differ

Yeah, I'm pretty sure that I didn't handle the boundary case correctly here. A shortcut is probably to check `floor(Int, nsteps) == 1`. But the final result seems incorrect...

Reply:
Also, I'm afraid we'll allocate with things like that `cur .+= δ; all(p .≤ cur .≤ P) || continue`. Using the current loop with primitive types is a guaranty for performance.

Reply:
Why would you want to pass `cur_x`, `cur_y`, `δx`, `δy`, `nsteps`, `length`, `step` when you can simply pass two ranges `rx, ry`? Writing a function with 10 arguments is indeed a bad code smell to me. Compared to better names and comments, splitting this into a function doesn't make it easier to decrypt.

> I'm afraid we'll allocate with things like that cur .+= δ; all(p .≤ cur .≤ P) || continue

I'm not sure if there will be worse. The whole purpose of StaticArrays is to store arrays as tuples so that Julia compiler can optimize the allocations away.

Reply:
`cur .+= δ` will be ok, `all(p .≤ cur .≤ P)` will allocate even if `cur` is a StaticArray, so this will be pretty bad. I'm rewriting this now with a while loop so that I can get rid of the runtime dispatch of the range arguments.

Reply:
I've found something way simpler, and removed the ugly helper functions.